File: 1018-binary-prefix-divisible-by-5.java
import java.util.*;

class BinaryPrefixDivisibleBy5 {
    public List<Boolean> prefixesDivBy5(int[] nums) {
        List<Boolean> res = new ArrayList<>();
        int cur = 0;
        for (int i = 0; i < nums.length; i++) {
            cur <<= 1;
            cur |= nums[i];
            cur %= 5; // we mod by 5 to prevent overflow
            res.add(cur % 5 == 0);
        }
        return res;
    }
  
    public static void main(String[] args) {
        BinaryPrefixDivisibleBy5 sol = new BinaryPrefixDivisibleBy5();
        System.out.println("Output: " + sol.prefixesDivBy5(new int[]{0, 1, 1}).toString() + ", expected: [true, false, false]");
        System.out.println("Output: " + sol.prefixesDivBy5(new int[]{1, 1, 1}).toString() + ", expected: [false, false, false]");
    }
}

File: 1018-binary-prefix-divisible-by-5.py
class Solution:
    def prefixesDivBy5(self, nums: list[int]) -> list[bool]:
        '''
        We have a loop that goes through each num, at each loop, we shift the bits by 1 to the left and OR with the num
        Next, get the result and mod 5 and store the result in a res array
        '''
        res = []
        cur = 0
        for num in nums:
            cur <<= 1
            cur |= num
            res.append(cur % 5 == 0)

        return res


sol = Solution()
print(
    f'output: {sol.prefixesDivBy5([0, 1, 1])}, expected: [True, False, False]')
print(
    f'output: {sol.prefixesDivBy5([1, 1, 1])}, expected: [False, False, False]')
print(
    f'output: {sol.prefixesDivBy5([0, 1, 1, 1, 1, 1])}, expected: [True, False, False, False, True, False]')
